src/resources/espnow.cc

```diff
@@ -0,0 +1,279 @@
+// Copyright (C) 2022 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "../top.h"
+
+#if defined(TOIT_FREERTOS) && defined(CONFIG_TOIT_ENABLE_ESPNOW)
+
+#include <esp_wifi.h>
+#include <esp_event.h>
+#include <esp_netif.h>
+#include <esp_now.h>
+#include <esp_log.h>
+
+#include "../objects_inline.h"
+#include "../process.h"
+#include "../resource.h"
+#include "../resource_pool.h"
+#include "../vm.h"
+
+#include "../event_sources/system_esp32.h"
+
+#define ESPNOW_RX_DATAGRAM_NUM      8
+#define ESPNOW_RX_DATAGRAM_LEN_MAX  250
+#define ESPNOW_TX_WAIT_US           1000
+
+namespace toit {
+
+struct DataGram {
+  bool used;
+  int len;
+  uint8_t mac[6];
+  uint8_t buffer[ESPNOW_RX_DATAGRAM_LEN_MAX];
+};
+
+const int kInvalidESPNow = -1;
+
+// Only allow one instance to use espnow.
+static  ResourcePool<int, kInvalidESPNow> espnow_pool(
+  0
+);
+
+static SemaphoreHandle_t tx_sem;
+static esp_now_send_status_t tx_status;
+static QueueHandle_t rx_queue;
+static SemaphoreHandle_t rx_datagrams_mutex;
+static struct DataGram* rx_datagrams;
+
+class ESPNowResourceGroup : public ResourceGroup {
+ public:
+  TAG(ESPNowResourceGroup);
+
+  ESPNowResourceGroup(Process* process);
+  ~ESPNowResourceGroup();
+};
+
+ESPNowResourceGroup::ESPNowResourceGroup(Process* process)
```

Comment:
The ResourceGroup is not created at boot time, it is created when the Toit program starts using ESPnow.  This means we have to guard against allocation failures.

It's probably better to have a trivial constructor in the ESPNowResourceGroup, then do the difficult work in the init primitive. That way, because you are in a primitive, you can return an error when allocations fail.

Reply:
Yes. Can the ESPNowResourceGroup's constructor function throw out an error when some errors trigger ?

Reply:
I know to implement according to discussion yesterday.